akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -974,3 +974,37 @@ private[stream] class DropWithin[T](timeout: FiniteDuration) extends SimpleLinea
 
   override def toString = "DropWithin"
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class Reduce[T](f: (T, T) ⇒ T) extends SimpleLinearGraphStage[T] {
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    override def toString = s"Reduce.Logic(aggregator=$aggregator)"
+    var aggregator: Option[T] = None
+
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        aggregator = Some(aggregator match {
+          case Some(agg) ⇒ f(agg, grab(in))
+          case None      ⇒ grab(in)
+        })
+        pull(in)
```

Comment:
shouldn't this be better performance via a `first` check or switch the `InHandler` after the first element arrived?